### ISSUE TITLE
Panics on drop will cause double drops.

### ISSUE THREAD

**Issue by Stebalien** [bluss/arrayvec#3]
IntoIter and ArrayVec need to be poisoned to prevent dropping before being drained; otherwise, if one of the elements panics on drop, it will be dropped twice. I've worked on a couple of solutions but they all have design tradeoffs so I figured I'd just let you fix it.

Drain doesn't have this issue because the ArrayVec's length is set to 0 when the Drain is initialized.

**Comment by Stebalien** [comment]
Test cases (abusing rust's panic on drop -> abort).
```rust
#[test]
#[should_panic]
fn test_drop_panic() {
    struct DropPanic;

    impl Drop for DropPanic {
        fn drop(&mut self) {
            panic!("drop");
        }
    }

    let mut array = ArrayVec::<[DropPanic; 1]>::new();
    array.push(DropPanic);
}

#[test]
#[should_panic]
fn test_drop_panic_into_iter() {
    struct DropPanic;

    impl Drop for DropPanic {
        fn drop(&mut self) {
            panic!("drop");
        }
    }

    let mut array = ArrayVec::<[DropPanic; 1]>::new();
    array.push(DropPanic);
    array.into_iter();
}
```

**Comment by bluss** [comment]
Thanks! I guess this won't be easy..

**Comment by bluss** [comment]
For the lols, please read #5. Hey you said you would leave it for me to fix! But I do think you have the sane version, and I the *insane*..

**Comment by bluss** [comment]
Fixed in version 0.3.10

**Comment by bluss** [comment]
Great, now it's fixed in smallvec too.
